### PR TITLE
Add recursion limit to inhabitedness check

### LINKED ISSUE #39489: Tricky type definition can cause inhabitedness check to go into an infinite loop.
This code:

```rust
struct Foo<'a, T: 'a> {
    ph: std::marker::PhantomData<T>,
    foo: &'a Foo<'a, (T, T)>,
}

fn wub(f: Foo<!>) {
    match f {}
}
```

Causes rustc to stack overflow. This should be pretty easy to "fix" by just putting a recursion limit in `is_uninhabited_from` and returning false when it gets hit. What's a suitable limit here? Presumably there are other limits like this in other places in the compiler.

### PR DESCRIPTION
**Pull request by canndrew**
Fixes #39489.
Add test aswell.

### CODE REVIEW

File: src/librustc/ty/inhabitedness/mod.rs [hunk 1]
```diff
@@ -145,8 +152,14 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
     pub fn uninhabited_from(
                 &self,
                 visited: &mut FxHashSet<(DefId, &'tcx Substs<'tcx>)>,
+                mut recursion_depth: u32,
                 tcx: TyCtxt<'a, 'gcx, 'tcx>) -> DefIdForest
     {
+        recursion_depth += 1;
+        if recursion_depth >= ARBITRARY_RECURSION_LIMIT {
```
Comment (arielb1) [inline src/librustc/ty/inhabitedness/mod.rs]: We normally use the `tcx.sess.recursion_limit` and emit a fatal error when it is exceeded (grep for that for usage examples).

**Comment by canndrew** [comment]
I've refactored the inhabitedness checking code a bit because there seemed to be a fairly narrow band of appropriate values for the recursion limit if it was implemented naively. `recursion_limit / 4` would cause perfectly sensible code hit the limit, `recursion_limit / 2` wouldn't kick in soon enough and the testcase would crash the compiler. Instead now we look at how deep we've recursed through a specific adt, if *that* number gets big then we're likely in an infinite loop.

**Comment by alexcrichton** [comment]
@bors: retry

* network error

**Pull request merged**